src/abt/generic.fun

```diff
@@ -0,0 +1,182 @@
+
+  fun @@ (f, x) = f x
+  infix @@ |> ||
+
+  structure Eff : MONAD =
+  struct
+    type 'a t = 'a generic
+    fun ret x = ([],[]) || x
+    fun bind f ((us', xs') || m) =
+      let
+        val (us, xs) || m' = f m
+      in
+        (us' @ us, xs' @ xs) || m'
+      end
+  end
+
+  structure G = struct structure F = MonadApplicative (Eff) open F Eff end
+
+  fun liftJdg {sort, subst} =
+    {sort = 
+      (fn (us, xs) || m => 
+         let 
+           val ((sigmas, taus), tau) = sort m
+           val (sigmas', taus') = (List.map #2 us, List.map #2 xs)
+         in
+           ((sigmas' @ sigmas, taus' @ taus), tau)
+         end),
+     subst = (fn env => G.map (subst env))}
+
+  fun map f (psi |> m) =
+    Tl.map (G.map f) psi |> m
+
+  fun ret isjdg jdg =
+    let
+      val x = L.fresh ()
+      val jdg' = G.ret jdg
+      val {sort,...} = liftJdg isjdg
+    in
+      Tl.singleton x jdg' |> L.var x (sort jdg')
+    end
+
+  structure Print = DebugShowAbt (L.Abt)
+
+  local
+    open L.Abt Tl.ConsView infix \ $#
+    structure Tl = TelescopeUtil (Tl)
+
+    fun substTele (isjdg : 'a isjdg) env (psi : 'a eff telescope) = 
+      Tl.map (G.map (#subst isjdg env)) psi
+
+    fun substState isjdg env (psi |> m) : 'a state =
+      substTele isjdg env psi |> L.subst env m
+
+    fun prependState psi' (psi |> abs) =
+      Tl.append psi' psi |> abs
+
+    fun prependBindings (symBindings, varBindings) abs = 
+      let
+        val (us', sigmas') = ListPair.unzip symBindings
+        val (xs', taus') = ListPair.unzip varBindings
+        val ((us, xs) \ m, ((sigmas, taus), tau)) = inferb abs
+        val binder = (us' @ us, xs' @ xs) \ m
+        val valence = ((sigmas' @ sigmas, taus' @ taus), tau)
+      in
+        checkb (binder, valence)
+      end
+
+    fun ::= (x, abs) = Metavar.Ctx.insert Metavar.Ctx.empty x abs
+    infix ::=
+
+    fun makeRebindingEnv (isjdg : 'a isjdg) bindings (psi : 'a telescope) =
+      let
+        val (symBindings, varBindings) = bindings
+        val (us', sigmas') = ListPair.unzip symBindings
+        val (xs', taus') = ListPair.unzip varBindings
+      in
+        Tl.foldl 
+          (fn (x, jdg, env) => 
+             let
+               val ((sigmas, taus), tau) = #sort isjdg jdg
+               val us = List.map (fn _ => Sym.named "u") sigmas
+               val xs = List.map (fn _ => Var.named "x") taus
+               (*check order of application here*)
+               val ps = ListPair.map (fn (u, sigma) => (O.P.ret u, sigma)) (us' @ us, sigmas' @ sigmas)
+               val ms = ListPair.map (fn (x, tau) => check (`x, tau)) (xs' @ xs, taus' @ taus)
+               val meta = check (x $# (ps, ms), tau)
+               val binder = (us, xs) \ meta
+               val abs = checkb (binder, ((sigmas, taus), tau))
+             in
+               Metavar.Ctx.insert env x abs
+             end)
+          Metavar.Ctx.empty
+          psi
+      end
+  in
+    fun 'a mul isjdg (ppsi |> abs) = 
```

Comment:
I would like to remark that it should be possible to replace this with a more efficient implementation, along the lines of `mul` for "pure" LCF w/o generic judgment. However, in order to get the damn thing even working, I wanted to keep the implementation as close to something I can reason about on paper as possible. Later on as it becomes better understood what is going on here, we will no doubt improve it.